stage0/src/acpi.rs

```diff
@@ -0,0 +1,411 @@
+                return Err("RSDP doesn't match expected size");
+            }
+
+            // Safety: we do not have concurrent threads so accessing the static is safe.
+            let buf = unsafe { RSDP.write(zeroed()) };
+
+            // Sanity checks.
+            if (buf.as_ptr() as *const _ as u64) < 0x80000
+                || (buf.as_ptr() as *const _ as u64) > 0x81000
+            {
+                log::error!("RSDP address: {:p}", buf);
+                return Err("RSDP address is not within the first 1 KiB of EBDA");
+            }
+            if (buf.as_ptr() as *const _ as u64) % self.align as u64 != 0 {
+                return Err("RSDP address not aligned properly");
+            }
+
+            fwcfg.read_file(&file, buf).map(|_| ())
+        } else if self.file() == ACPI_TABLES_FILE_NAME {
+            // Safety: we do not have concurrent threads so accessing the static is safe.
+            let buf = unsafe { EBDA.write(zeroed()) };
+
+            if (buf.as_ptr() as *const _ as u64) % self.align as u64 != 0 {
+                log::error!(
+                    "ACPI tables address: {:p}, required alignment: {}",
+                    buf,
+                    self.align
+                );
+                return Err("ACPI tables address not aligned properly");
+            }
+            fwcfg.read_file(&file, buf).map(|_| ())
+        } else {
+            Err("Unsupported file in table-loader")
+        }
+    }
+}
+
+impl Debug for Allocate {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.debug_struct("Allocate")
+            .field("file", &self.file())
+            .field("align", &self.align)
+            .field("zone", &self.zone())
+            .finish()
+    }
+}
+
+/// COMMAND_ADD_POINTER - patch the table (originating from `dest_file`) at `offset`, by adding a
+/// pointer to the table originating from `src_file`.
+///
+/// 1,2,4 or 8 byte unsigned addition is used depending on `size`.
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct AddPointer {
+    dest_file: RomfileName,
+    src_file: RomfileName,
+    offset: u32,
+    size: u8,
+    _padding: [u8; 6],
+}
+static_assertions::assert_eq_size!(AddPointer, Pad);
+
+impl AddPointer {
+    pub fn dest_file(&self) -> &CStr {
+        CStr::from_bytes_until_nul(&self.dest_file).unwrap()
+    }
+
+    pub fn src_file(&self) -> &CStr {
+        CStr::from_bytes_until_nul(&self.src_file).unwrap()
+    }
+
+    fn invoke(&self) -> Result<(), &'static str> {
+        let dest_file = get_file(self.dest_file())?;
+        let src_file = get_file(self.src_file())?;
+
+        if self.offset as usize + self.size as usize > dest_file.len() {
+            return Err("Write for COMMAND_ADD_POINTER would overflow destination file");
+        }
+        if self.size > 8 || self.size.count_ones() != 1 {
+            return Err("COMMAND_ADD_POINTER has invalid size");
+        }
+        let mut pointer = 0u64;
+        unsafe {
```

Comment:
Oh, I forgot to mention: I think these unsafe blocks can be removed by using something like:

`pointer.as_bytes_mut()[..self.size].copy_from_slice(&dest_file[self.offset..(self.offset+size)])`

and

`dest_file[self.offset..(self.offset+size)].copy_from_slice(pointer.as_bytes()[..self.size])`

Reply:
Much cleaner, thanks!